runtime(termdebug): Allow remote debugging over tty (ssh, wsl, docker ...) (#18429)
vim/vim: runtime/doc/terminal.txt

"continue" being used for the `:Continue` command, instead of using the
communication channel.

Remote Debugging ~
							*termdebug-remote*
One of the main issues of remote debugging is the access to the debuggee's
source files. The plugin can profit from system and vim's networking
capabilities to workaround this.
						*termdebug-substitute-path*
Use the `g:termdebug_config['substitute_path']` entry to map remote to local

> files using the same strategy that gdb's `substitute-path` command.

Reviewer: ```suggestion
files using the same strategy that gdb's `substitute-path` command uses.
```

For example:
- Use |netrw| to access files remoting via ssh: >
    let g:termdebug_config['command'] = ['ssh', 'hostname', 'gdb']
    let g:termdebug_config['substitute_path'] = { '/': 'scp://hostname//' }

> < Note: that first is specified the remote machine root path and later

Reviewer: ```suggestion
< Note: that the key specifies the remote machine root path and the value
```

  the local one.
- Use Windows' `UNC` paths to access `WSL2` sources: >
    let g:termdebug_config['command'] = ['wsl', 'gdb']
    let g:termdebug_config['substitute_path'] = {
	\ '/': '\\wsl.localhost\Ubuntu-22.04\',
	\ '/mnt/c/': 'C:/' }
< Note: that several mappings are required: one for each drive unit

>   and one for the linux filesystem (queried via `wslpath`).

Reviewer: Do we always need to specify `g:termdebug_config['command']` and also `g:termdebug_config['substitute_path']` ?

Can you please also give an example. Something along the lines of:
```
Assuming you want to debug a program on a remote server using ssh, you would need to run the following command: >
ssh user@hostname gdb args /opt/command

This needs to be translated into a command your local Vim can run: >
gdb args scp://user@hostname//opt/command
```

Author: I have reproduced the session example as a remote session.

The plugin "program window" is linked to `gdb` using `pty slave devices`.
Those must be local to the `gdb` instance, thus remote debugging only works
on `prompt mode` (see |termdebug_use_prompt|).
In this mode any `ssh` or `wsl` command would be detected and a similar command
used to launch a remote `tty` terminal session and connect it to `gdb`.

						*termdebug-remote-window*

> In order to use another remote terminal clients, set "remote_window" entry

Reviewer: ```suggestion
In order to use another remote terminal client, set "remote_window" entry
```

> In order to use another remote terminal clients, set "remote_window" entry

Author: If I were referencing a single client, e.g. `telnet`, no extra option would be required. I would have included extra builtin management (like with `ssh` or `wsl`) 🤣.

communication channel.

Remote debugging ~
							*termdebug-remote*
One of the main issues of remote debugging is the access to the debuggee's
source files. The plugin can profit from system and vim's networking
capabilities to workaround this.
						*termdebug-remote-example*
The |termdebug-example| can be replicated by running the `gdb` debugger and the

> Vim debuggee on a remote Linux machine accessible via `ssh`.

Reviewer: I found debuggee a bit unusual, so perhaps:

```suggestion
The |termdebug-example| can be replicated by running the `gdb` debugger to
to debug Vim on a remote Linux machine accessible via `ssh`.
```

Reviewer 2: ... but not ``to to``

- Build Vim as explained in the local example.

> - If 'socat' is not available in the remote machine 'terminal' mode will not

Reviewer: ```suggestion
- If "socat" is not available in the remote machine 'terminal' mode will not
```

						*termdebug-substitute-path*
Use the `g:termdebug_config['substitute_path']` entry to map remote to local
files using the same strategy that gdb's `substitute-path` command uses.
For example:
- Use |netrw| to access files remoting via ssh: >
    let g:termdebug_config['command'] = ['ssh', 'hostname', 'gdb']
    let g:termdebug_config['substitute_path'] = { '/': 'scp://hostname//' }
< Note: that the key specifies the remote machine root path and the value
  the local one.
- Use Windows' `UNC` paths to access `WSL2` sources: >
    let g:termdebug_config['command'] = ['wsl', 'gdb']
    let g:termdebug_config['substitute_path'] = {
	\ '/': '\\wsl.localhost\Ubuntu-22.04\',
	\ '/mnt/c/': 'C:/' }
< Note: that several mappings are required: one for each drive unit
  and one for the linux filesystem (queried via `wslpath`).

In this mode any `ssh` or `wsl` command would be detected and a similar
command would be used to launch `socat` in a remote `tty` terminal session
and connect it to `gdb`.
If `socat` is not available a plain remote terminal would be used as
fallback.
The next session shows how to override this default behaviour.

						*termdebug-remote-window*

> In order to use another remote terminal client, set 'remote_window' entry

Reviewer: ```suggestion
In order to use another remote terminal client, set "remote_window" entry
```

in `g:termdebug_config` variable before invoking `:Termdebug`. For example:

> - Debugging inside a docker container on 'prompt' mode: >

Reviewer: ```suggestion
- Debugging inside a docker container using "prompt" mode: >
```

    let g:termdebug_config['use_prompt'] = v:true
    let g:termdebug_config['command'] = ['docker', 'run', '-i',
	\ '--rm', '--name', 'container-name', 'image-name', 'gdb']
    let g:termdebug_config['remote_window'] =
	\ ['docker', 'exec', '-ti', 'container-name'
	\ ,'socat', '-dd', '-', 'PTY,raw,echo=0']

> - Debugging inside a docker container on a 'terminal buffer'.

Reviewer: ```suggestion
- Debugging inside a docker container using a "terminal buffer".
```

  The container should be already running because unlike the previous

>   case for `terminal mode` 'program' and 'communication' ptys are created

Reviewer: ```suggestion
  case for `terminal mode` "program" and "communication" ptys are created
```

  before the gdb one: >
    $ docker run -ti --rm --name container-name immage-name
<  Then, launch the debugger: >
    let g:termdebug_config['use_prompt'] = v:false " default
    let g:termdebug_config['command'] =
	\ ['docker', 'exec', '-ti', 'container-name', 'gdb']
    let g:termdebug_config['remote_window'] =
	\ ['docker', 'exec', '-ti', 'container-name'
	\ ,'socat', '-dd', '-', 'PTY,raw,echo=0']

> Note: 'command' cannot use `-t` on |termdebug-prompt| mode because prompt

Reviewer: ```suggestion
Note: "command" cannot use `-t` on |termdebug-prompt| mode because prompt
```

buffers cannot handle `tty` connections.
The 'remote_window' command must use `-t` because otherwise it will lack
a `pty slave device` for gdb to connect.

> Note: 'socat' must be available in the remote machine on 'terminal' mode.

Reviewer: ```suggestion
Note: "socat" must be available in the remote machine on "terminal" mode.
```

If you want to modify them, add a function to filter the argument list: >
	let g:termdebug_config['command_filter'] = MyDebugFilter

> A 'command_filter' scenario is solving escaping issues on remote debugging

Reviewer: ```suggestion
A "command_filter" scenario is solving escaping issues on remote debugging
```

> over 'ssh'. For convenience a default filter is provided for escaping

Reviewer: ```suggestion
over "ssh". For convenience a default filter is provided for escaping
```

> whitespaces inside the arguments. It is automatically configured for 'ssh',

Reviewer: ```suggestion
whitespaces inside the arguments. It is automatically configured for "ssh",
```
